src/cpio.rs

```diff
@@ -0,0 +1,160 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! cpio miniroot support.
+
+use crate::io;
+use crate::ramdisk;
+use crate::result::{Error, Result};
+use crate::{print, println};
+use alloc::boxed::Box;
+use core::slice;
+
+pub(crate) struct FileSystem {
+    sd: io::Sd,
+}
+
+impl FileSystem {
+    pub(crate) fn try_new(bs: &[u8]) -> Result<FileSystem> {
+        if bs.starts_with(b"070707") {
+            let sd = io::Sd::from_slice(bs);
+            Ok(FileSystem { sd })
+        } else {
+            Err(Error::FsInvMagic)
+        }
+    }
+}
+
+pub(crate) struct File {
+    data: io::Sd,
+}
+
+impl File {
+    fn as_slice(&self) -> &[u8] {
+        let ptr = self.data.as_ptr();
+        let len = self.data.len();
+        unsafe { slice::from_raw_parts(ptr, len) }
+    }
+}
+
+impl ramdisk::File for File {
+    fn file_type(&self) -> ramdisk::FileType {
+        ramdisk::FileType::Regular
+    }
+}
+
+impl io::Read for File {
+    fn read(&self, offset: u64, dst: &mut [u8]) -> Result<usize> {
+        let s = self.as_slice();
+        s.read(offset, dst)
+    }
+
+    fn size(&self) -> usize {
+        self.data.len()
+    }
+}
+
+impl ramdisk::FileSystem for FileSystem {
+    fn open(&self, path: &str) -> Result<Box<dyn ramdisk::File>> {
+        let ptr: *const u8 = self.sd.as_ptr();
+        let len = self.sd.len();
+        let cpio = unsafe { core::slice::from_raw_parts(ptr, len) };
+        let key = path.strip_prefix("/").unwrap_or(path);
+        for file in cpio_reader::iter_files(cpio) {
+            if file.name() == key {
+                let data = io::Sd::from_slice(file.file());
+                return Ok(Box::new(File { data }));
+            }
+        }
+        Err(Error::FsNoFile)
+    }
+
+    fn list(&self, path: &str) -> Result<()> {
+        let ptr: *const u8 = self.sd.as_ptr();
+        let len = self.sd.len();
+        let cpio = unsafe { core::slice::from_raw_parts(ptr, len) };
+        let key = path.strip_prefix('/').unwrap_or(path);
+        for file in cpio_reader::iter_files(cpio) {
+            if file.name() == key {
+                lsfile(path, &file);
+                return Ok(());
+            }
+        }
+        let mut found = false;
+        for file in cpio_reader::iter_files(cpio) {
+            if file.name().starts_with(key) {
+                lsfile(file.name(), &file);
+                found = true;
+            }
+        }
+        if found { Ok(()) } else { Err(Error::FsNoFile) }
+    }
+
+    fn as_str(&self) -> &str {
+        "cpio"
+    }
+
+    fn with_addr(&self, addr: usize) -> *const u8 {
+        self.sd.as_ptr().with_addr(addr)
```

Comment:
i do wonder if `.as_ptr()` should just be called `ptr()`, or `data()`. `sd.as_ptr()` sure reads like we're getting a `*Sd` at some point..

Reply:
Done: Changed to `.data`.